rover_video/src/camera_node.cpp

```diff
@@ -0,0 +1,558 @@
+                                        }));
+
+        if (!_videoWatchDog.joinable())
+        {
+            StartWatchDog();
+        }
+
+        // Access the recording using at() to safely get the reference
+        Recording* pRecording = &_RecordingMap.at(cameraURL_);
+
+        if (pRecording->startRecording())
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+}
+
+/**
+ * @brief Writes the current value of the GPS onto placeholding variables
+ *
+ * @param gps_message_ Address reference of the GPS subscriber
+ */
+void CameraNode::callbackPosition(const rover_msgs::msg::GpsPosition& gps_message_)
+{
+    _last_latitude = gps_message_.latitude;
+    _last_longitude = gps_message_.longitude;
+}
+
+/**
+ * @brief Add the URL to the shutdown list and notify watchdog to process shutdown
+ *        This function is passed as a callback function to the recording class
+ * @param camURL_ key for the hashmap
+ */
+void CameraNode::RequestShutdown(std::string camURL_)
+{
+    RCLCPP_WARN(LOGGER, "Received shutdown request for %s", camURL_.c_str());
+
+    {
+        std::unique_lock<std::mutex> lock(_recordingMutex);
+        _RecordingShutdownRequestSet.insert(camURL_);
+    }  // unlock
+    _recordingCv.notify_one();
+    return;
+}
+
+/**
+ * @brief Start the Watchdog thread
+ *
+ */
+bool CameraNode::StartWatchDog(void)
+{
+    _watchDogStop.store(false);
+    _videoWatchDog = std::thread(&CameraNode::VideoWatchDogFunction, this);
+    return true;
+}
+
+/**
+ * @brief Shutdown and erase a recording from the hashmap
+ * @until watchDogStop = true
+ */
+void CameraNode::VideoWatchDogFunction(void)
+{
+    RCLCPP_DEBUG(LOGGER, "Starting video watchdog");
+    while (!_watchDogStop.load())
+    {
+        std::unique_lock<std::mutex> lock(_recordingMutex);
+        _recordingCv.wait(lock,
+                          [this]
+                          {
+                              return _watchDogStop.load() || !_RecordingShutdownRequestSet.empty();
+                          });
+
+        if (_watchDogStop)
+            break;
+        else
+        {
+            for (std::string url : _RecordingShutdownRequestSet)
+            {
+                RCLCPP_WARN(LOGGER, "Processing Shutdown for %s", url.c_str());
+                if (_RecordingMap.find(url) != _RecordingMap.end())
+                {
+                    if (!_RecordingMap.erase(url))
+                    {
+                        RCLCPP_ERROR(LOGGER, "Shutdown request for %s could not be processed, please try again", url.c_str());
+                    }
+
+                    if (_RecordingMap.empty())
+                    {
+                        _watchDogStop.store(true);
+                    }
+                }
+                else
+                {
+                    RCLCPP_ERROR(LOGGER, "Unable to find %s for shutdown, please try again", url.c_str());
```

Comment:
Be a bit more specific, for example like this:

```suggestion
                    RCLCPP_WARN(LOGGER, "No recording on %s, no action done", url.c_str());
```